### ISSUE TITLE
Different provider.layer depending on zoom level (user defined layer names)

### ISSUE THREAD

**Issue by thomersch** [terranodo/tegola#94]
(Somehow related to #87 and #81)

I use generalised tables in order to reduce the size of tiles, e.g. "osm_roads" contains full resolution geometries, while those in "osm_roads_gen0" are reduced.

The provider layer configuration looks like this:

	[[providers.layers]]
 	name = "roads"
 	geometry_fieldname = "geom"
 	id_fieldname = "osm_id"
 	sql = """
  		SELECT name, osm_id, class, type, ST_AsBinary(geometry) AS geom
  		FROM osm_roads
  		WHERE geometry && !BBOX!
  	"""

  	[[providers.layers]]
 	name = "roads_gen0"
 	geometry_fieldname = "geom"
 	id_fieldname = "osm_id"
 	sql = """
  		SELECT name, osm_id, class, type, ST_AsBinary(geometry) AS geom
  		FROM osm_roads_gen0
  		WHERE geometry && !BBOX!
  	"""

My maps config looks like this:

	[[maps]]
	name = "foobar"

	[[ maps.layers ]]
	provider_layer = "osm.roads"
	min_zoom = 11
	max_zoom = 14

	[[ maps.layers ]]
	provider_layer = "osm.roads_gen0"
	min_zoom = 8
	max_zoom = 10

Now, I would like to be able to use the roads_gen0 layer on the roads layer. This could be solved by being able to specify a layer output name in the maps.layers config.

**Comment by ARolek** [comment]
@thomersch when you say:

>  I would like to be able to use the roads_gen0 layer on the roads layer

Are you referring to having both "osm.roads" and "osm.roads_gen0" render at the same zoom (i.e. zoom of 10)? If so, that's just a minor change to the "osm.raods" `min_zoom` level. If not, could you give me a few more details on your issue?

**Comment by thomersch** [comment]
Sorry for being confusing. I will try to do better :)

What I mean is: Up to zoom level 10 I want to have `osm.roads_gen0` as data source for the roads layer, but on all layers above (> 10) I want to have `osm.roads` as data source for the same roads layer.

Same layer, same rendering in the end, but with data from different tables.

The background is, as I mentioned, that `roads_gen0` has low resolution data which is good enough for low zoom levels (up to 10) while `roads` has all the data for the high levels.

**Comment by ARolek** [comment]
Got it. Your config looks good. Are both tables in the same database? What results are you seeing?

**Comment by thomersch** [comment]
Yes, they are both in the same DB.

Well, I get two layers: One called `roads` and the other called `roads_gen0`, which is troublesome, because I want to apply the exact same style to them.

One possibility to solve it, would be the ability to give aliases to `maps.layers`. I could name them both "roads" and depending on the zoom level I would get the right thing. Admittedly kind of quick'n'dirty.
Another option would be: Add the ability to specify different queries depending on the zoom level. But this doesn't fit the configuration structure very well.

**Comment by ARolek** [comment]
I see what you're trying to do, and allowing for an explicit name in the `map.layers` is an interesting way to solve it. Can you provide a snippet from your style config? I would like to see your styling approach.

One alternative approach that you can try now is to use [Map Layer Default Tags](http://tegola.io/configuration/#map-layer-default-tags) in the config to give both road layers a `class` or another key / value pair for your style sheet to key off of.

**Comment by thomersch** [comment]
I use mapboxgl.js for rendering. My source definition looks like this:

```json
"sources": {
	"osm": {
		"tiles": ["http://localhost:8123/foobar/{z}/{x}/{y}.pbf"],
		…
	}
}
```

Later on I pick up the source in multiple layers, e.g.

```json
{
	"id": "roads-secondary",
	"type": "line",
	"source": "osm",
	"source-layer": "roads",
	"paint": {
		"line-color": "#FFE873",
		"line-width": {
			"base": 1,
			"stops": [[8, 0], [12, 3], [16, 8]]
		}
	},
	"filter": ["==", "type", "secondary"]
}
```

```json
{
	"id": "roads-primary",
	"type": "line",
	"source": "osm",
	"source-layer": "roads",
	"paint": {
		"line-color": "#FFC859",
		"line-width": {
			"base": 1,
			"stops": [[9, 1], [13, 5], [16, 10]]
		}
	},
	"layout": {
		"line-cap": "round",
		"line-join": "bevel"
	},
	"filter": ["in", "type", "primary", "trunk"]
}
```

By the way, my gen0 table is filtered, so I don't have to transmit all the ways: `"sql_filter": "type IN ('motorway', 'primary', 'secondary')`.

All of this is still in development stage, so possibly I will have to filter more aggressively, filter more or restyle the whole thing.

> One alternative approach that you can try now is to use Map Layer Default Tags in the config to give both road layers a class or another key / value pair for your style sheet to key off of.

Good point. But as far as I see this is not unproblematic, because I have to select a layer first before filtering.

**Comment by ARolek** [comment]
Ok this is what I figured was going on but I wanted to be sure. I need to think how supporting a user defined layer name will impact other parts of the software. This is a valid use case and something I'm sure other's will run into. I will do some research and circle back with some possibilities.

**Comment by thomersch** [comment]
Today I started hacking on the code base, maybe I'll be able to come up with something.

**Comment by ARolek** [comment]
I have been doing some more research on this issue while reviewing your pull request and the approach here is essentially "VT compositing". We have discussed compositing layers from remote vector tile sources before (essentially a remote data provider) which would allow for using tegola to "mash-up" data from other vector tile servers and combine the data into a single tile. There's a quick post about some of the naming issues that arise with this approach: [VT compositing and named layers](https://gist.github.com/springmeyer/3522514c85443b57e338).

The following two threads revolve around sharing client side styling. The discussions have been open for awhile. It might be worth jumping in to see where the conversation stands:
- [#224: Support using multiple "source layers" in one "style layer"](https://github.com/mapbox/mapbox-gl-style-spec/issues/224)
- [#231: Decoupling styling structure from source structure](https://github.com/mapbox/mapbox-gl-style-spec/issues/231)

For reference, the [MVT Layer Spec](https://github.com/mapbox/vector-tile-spec/tree/master/2.1#41-layers) for reference.

I have a few other points that I will bring up on the PR.

**Comment by pnorman** [comment]
This seems to be a duplicate of #87, except covering layers coming from the same provider.

**Comment by ARolek** [comment]
@dracic I think you're looking for our !ZOOM! token which is available in [v0.4.0-alpha](https://github.com/terranodo/tegola/releases/tag/v0.4.0-alpha).

**Comment by dracic** [comment]
@ARolek, thank you for your info. That's exactly what I was looking for...

**Comment by ARolek** [comment]
@dracic perfect! let me know if you have any issues while testing the alpha version. We have a couple more clipping edge cases being address in `issue-56` but should be cutting beta soon.

**Comment by ARolek** [comment]
This has been implemented in v0.4.0 pre release builds which can be found [here](https://github.com/terranodo/tegola/releases)
